sdks/python/apache_beam/transforms/util.py

```diff
@@ -712,10 +824,20 @@ def __init__(
         clock=clock,
         record_metrics=record_metrics)
     self._element_size_fn = element_size_fn
+    self._max_batch_dur = max_batch_duration_secs
+    self._clock = clock
 
   def expand(self, pcoll):
     if getattr(pcoll.pipeline.runner, 'is_streaming', False):
       raise NotImplementedError("Requires stateful processing (BEAM-2687)")
+    elif self._max_batch_dur is not None:
+      coder = coders.registry.get_coder(pcoll)
+      return pcoll | WithKeys(0) | ParDo(
```

Comment:
Non-blocking for this PR, but something we may want to consider; rather than using a single fixed key, does it make sense to try to have a single key per worker somehow? (one way to do this would be using multi_process_shared.py)

That way we're still batching per machine in a parallelizable way, but we get stateful batching across bundles.

The current implementation is likely still useful for many use cases where batching is not the expensive part (e.g. RunInference) or there are few workers.

Reply:
At one point I suggested keying with worker IDs, it may be worth coming back to that idea

Reply:
Oh I missed that (or forgot), hopefully I wasn't against it initially?

Regardless, I like having this as our first pass, we can see how it performs and go from there